internal/podinjector/main.go

```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+func (p *podSidecarInjector) InjectDecoder(d *admission.Decoder) error {
+	p.decoder = d
+	return nil
+}
+
+func (p *podSidecarInjector) mutate(ctx context.Context, ns corev1.Namespace, pod corev1.Pod) (corev1.Pod, error) {
+	logger := p.logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
+
+	// if no annotations are found at all, just return the same pod
+	annValue := sidecar.AnnotationValue(ns, pod)
+	if len(annValue) == 0 {
+		logger.V(1).Info("annotation not present in deployment, skipping sidecar injection")
+		return pod, nil
+	}
+
+	// is the annotation value 'false'? if so, we need a pod without the sidecar (ie, remove if exists)
+	if strings.EqualFold(annValue, "false") {
+		logger.V(1).Info("pod explicitly refuses sidecar injection, attempting to remove sidecar if it exists")
+		return sidecar.Remove(pod)
+	}
+
+	// from this point and on, a sidecar is wanted
+
+	// check whether there's a sidecar already -- return the same pod if that's the case.
+	if sidecar.ExistsIn(pod) {
+		logger.V(1).Info("pod already has sidecar in it, skipping injection")
+		return pod, nil
+	}
+
+	// which instance should it talk to?
+	otelcol, err := p.getCollectorInstance(ctx, ns, annValue)
+	if err != nil {
+		if err == ErrMultipleInstancesPossible || err == ErrNoInstancesAvailable || err == ErrInstanceNotSidecar {
+			// we still allow the pod to be created, but we log a message to the operator's logs
+			logger.Error(err, "failed to select an OpenTelemetry Collector instance for this pod's sidecar")
+			return pod, nil
+		}
+
+		// something else happened, better fail here
+		return pod, err
+	}
+
+	// once it's been determined that a sidecar is desired, none exists yet, and we know which instance it should talk to,
+	// we should add the sidecar.
+	logger.V(1).Info("injecting sidecar into pod", "otelcol-namespace", otelcol.Namespace, "otelcol-name", otelcol.Name)
+	return sidecar.Add(p.config, p.logger, otelcol, pod)
+}
+
+func (p *podSidecarInjector) getCollectorInstance(ctx context.Context, ns corev1.Namespace, ann string) (v1alpha1.OpenTelemetryCollector, error) {
+	if strings.EqualFold(ann, "true") {
+		return p.selectCollectorInstance(ctx, ns, ann)
+	}
+
+	otelcol := v1alpha1.OpenTelemetryCollector{}
+	err := p.client.Get(ctx, types.NamespacedName{Name: ann, Namespace: ns.Name}, &otelcol)
+	if err != nil {
+		return otelcol, err
+	}
+
+	if otelcol.Spec.Mode != v1alpha1.ModeSidecar {
+		return v1alpha1.OpenTelemetryCollector{}, ErrInstanceNotSidecar
+	}
+
+	return otelcol, nil
+}
+
+func (p *podSidecarInjector) selectCollectorInstance(ctx context.Context, ns corev1.Namespace, ann string) (v1alpha1.OpenTelemetryCollector, error) {
+	otelcols := v1alpha1.OpenTelemetryCollectorList{}
+	if err := p.client.List(ctx, &otelcols, client.InNamespace(ns.Name)); err != nil {
```

Comment:
Is this only looking for collectors in the same namespace? Is not possible to have the collector on another namespace and use true as an annotation value ?

Reply:
For now, yes. @objectiser gave a nice idea about adding a list of namespaces to the CR for Mode=Sidecar, so that we could have all otelcols in the same namespace, each taking care of a set of namespaces. But I still have to figure out RBAC for that, so, it might come only in a future PR.